pkg/imgutil/dockerconfigresolver/registryurl.go

```diff
@@ -0,0 +1,147 @@
+   limitations under the License.
+*/
+
+package dockerconfigresolver
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+
+	"github.com/containerd/log"
+)
+
+// Parse will return a normalized Docker Registry url from the provided string address
+func Parse(address string) (*RegistryURL, error) {
+	log.L.Debugf("parsing docker registry URL %q", address)
+	var err error
+	// No address or address as docker.io? Default to standardized index
+	if address == "" || address == "docker.io" {
+		log.L.Debugf("normalized to %q", dockerIndexServer)
+		address = dockerIndexServer
+	}
+	// If it has no scheme, slap one just so we can parse
+	if !strings.Contains(address, "://") {
+		address = fmt.Sprintf("%s://%s", schemeHTTPS, address)
+	}
+	// Parse it
+	u, err := url.Parse(address)
+	if err != nil {
+		log.L.Debug("unparsable - giving up")
+		return nil, errors.Join(ErrUnparsableURL, err)
+	}
+	sch := scheme(u.Scheme)
+	// Scheme is entirely disregarded anyhow, so, just drop it all and set to https
+	if sch == schemeHTTP {
+		log.L.Debug("changing http to https")
+		u.Scheme = string(schemeHTTPS)
+	} else if sch != schemeHTTPS && sch != schemeNerdctlExperimental {
+		log.L.Debugf("unrecognized scheme %q", sch)
+		// Docker is wildly buggy when it comes to non-http schemes. Being more defensive.
+		return nil, ErrUnsupportedScheme
+	}
+	// If it has no port, add the standard port explicitly
+	if u.Port() == "" {
+		log.L.Debug("adding standard port")
+		u.Host = u.Hostname() + ":" + standardHTTPSPort
+	}
+	reg := &RegistryURL{URL: *u}
+	queryParams := u.Query()
+	nsQuery := queryParams.Get(namespaceQueryParameter)
+	if nsQuery != "" {
+		log.L.Debugf("this is a namespaced url, parsing namespace %q", nsQuery)
+		reg.Namespace, err = Parse(nsQuery)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return reg, nil
+}
+
+// RegistryURL is a struct that represents a registry namespace or host, meant specifically to deal with
+// credentials storage and retrieval inside Docker config file.
+type RegistryURL struct {
+	url.URL
+	Namespace *RegistryURL
+}
+
+// CanonicalIdentifier returns the identifier expected to be used to save credentials to docker auth config
+func (rn *RegistryURL) CanonicalIdentifier() string {
+	log.L.Debugf("retrieving canonical identifier for %q", rn.URL.String())
+	// If it is the docker index over https, port 443, on the /v1/ path, we use the docker fully qualified identifier
+	if rn.Scheme == string(schemeHTTPS) && rn.Hostname() == "index.docker.io" && rn.Path == "/v1/" && rn.Port() == standardHTTPSPort ||
+		rn.URL.String() == dockerIndexServer {
+		log.L.Debugf("assimilated to docker %q", dockerIndexServer)
+		return dockerIndexServer
+	}
+	// Otherwise, for anything else, we use the hostname+port part
+	identifier := rn.Host
+	// If this is a namespaced entry, wrap it, and slap the path as well, as hosts are allowed to be non-compliant
+	if rn.Namespace != nil {
+		log.L.Debug("namespaced identifier")
```

Comment:
Do we still need these debug logs? Maybe too verbose.

Reply:
https://github.com/containerd/nerdctl/pull/3293#discussion_r1713121552

We can remove most of them.
Let me do that asap.